### PR TITLE
[Verif] Add contract examples to dialect doc

### PR DESCRIPTION
**Pull request by fabianschuiki**
Add a few examples of how contracts can be used to modularize formal verification. These will likely become integration tests once the machinery to execute them is in place.

### CODE REVIEW

**Review by maerhart — COMMENTED** [review]
Really nice walkthrough of how contracts work! 🎉 
Cool how all these example contracts could also be lowered to `verif.lec` and proven natively in CIRCT 🙂

File: docs/Dialects/Verif.md [hunk 1]
```diff
@@ -4,6 +4,334 @@ This dialect provides a collection of operations to express various verification
...
+  // Compute 9*a as (a<<3)+a.
+  %c3_i42 = hw.constant 3 : i42
+  %0 = comb.shl %a, %c3_i42 : i42
+  %1 = comb.add %a, %0 : i42
+
+  // Contract to check that the circuit actually produces 9*a.
+  %z = verif.contract %1 : i42 {
+    %c9_i42 = hw.constant 9 : i42
+    %a9 = comb.mul %a, %c9_i42 : i42
+    verif.ensure_equal %z, %a9
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply docs/Dialects/Verif.md:42]: You could definitely create an equality assertion implicitly as part of the lowering. This would prevent you from providing symbolic or more general replacement values though. In the carry-save adder compressor example further down, the contract creates symbolic replacement values that are constrained to sum up to a given value. In that example you want to make the contract hide the exact value of the output as an implementation detail.

[the remark this answers is not in the document]
Comment (fabianschuiki) [reply docs/Dialects/Verif.md:42]: Using the `%z` in the `ensure_equal` instead of the `%1` is important because it allows the original implementation to become unused very easily when the contract is assumed to hold. Otherwise you'd be left with a `verif.assume_equal %1, %a9`, which still uses `%1`. It's not entirely trivial to figure out if that assume could be removed or not :thinking:. For example, the assume could contrain an input further than the contract, and so it might not be safe to get rid of it. Actually making the contract put assurances on the contract output value instead of specifically the implementation fixes that.

File: docs/Dialects/Verif.md [hunk 2]
```diff
@@ -4,6 +4,334 @@ This dialect provides a collection of operations to express various verification
...
+  }
+
+  hw.output %z : i42
+}
+```
+
+Leveraging the contract in this module to simplify verification requires two steps: proving that the contract holds, and then assuming that it holds everywhere where it's used.
+Proving the contract can be done by pulling it and all its fan-in cone out into a `verif.formal` op.
+There it can be inlined by passing its operands through to its results and replacing ensure with assert and require with assume.
+In a nutshell, passing through its operands to the results and checking the ensures verifies that the ensures hold assuming that the contract isn't there and instead the actual hardware implementation is used.
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply docs/Dialects/Verif.md:53]: Yeah this is a pretty horrible sentence :sweat_smile:. Let me rephrase that.

File: docs/Dialects/Verif.md [hunk 3]
```diff
@@ -4,6 +4,334 @@ This dialect provides a collection of operations to express various verification
...
+
+hw.module @CarrySaveAdder5_AssumeContract_Simplified(
+  in %a0: i42, in %a1: i42, in %a2: i42, in %a3: i42, in %a4: i42,
+  out z: i42
+) {
+  %inputSum = comb.add %a0, %a1, %a2, %a3, %a4, %a5 : i42
+  hw.output %inputSum : i42
+}
+```
+
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply docs/Dialects/Verif.md:334]: :grin: Yeah good point. Maybe some one-hot decoder that only works when the input is truly one-hot?

File: docs/Dialects/Verif.md [hunk 4]
```diff
@@ -4,6 +4,334 @@ This dialect provides a collection of operations to express various verification
...
+    // We don't want this contract to give guarantees about what the exact
+    // values of its outputs are going to be. Instead, we only want to guarantee
+    // that they sum up to the right number. To express this, we pick two
+    // symbolic values and constrain them to sum up to that number, and yield
+    // those from the contract. This says "you'll get any two numbers that sum
+    // up to the sum of the inputs".
+    %any0 = verif.symbolic_value : i42
+    %any1 = verif.symbolic_value : i42
+    %anySum = comb.add %any0, %any1 : i42
+    verif.assume_equal %anySum, %inputSum : i42
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply docs/Dialects/Verif.md:157]: Hmm, that's a very interesting thought. Maybe these two things are separate. The idea behind the contracts was to automate the process of assert-assume swapping, to ensure that you're actually only assuming what you've asserted before. I wonder if the substitutions are an orthogonal piece of the puzzle :thinking: How would you split the above up into a contract and a potentially new op that encodes substitution?

File: docs/Dialects/Verif.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (maerhart) [reply docs/Dialects/Verif.md:42]: > This would prevent you from providing symbolic or more general replacement values though.

Good point! In that case you'd want the `ensure_equal` to check the addition of the outputs for equality, which you can't know when inserting them implicitly. But this is also solved by the comment further down about applying the contract to the alternative implementation as well (i.e., the value you return here).

> It's not entirely trivial to figure out if that assume could be removed or not

Right, that's actually a genius way to simplify the implementation here!
If you have the `verif.assume_equal %1, %a9`, the contract already verified that they are the same. So you can just pick whichever you prefer (i.e., you can replace all uses of `%1` with `%a9` during inlining). But, it's an additional step during inlining and thus a bit more work.
Just looking at this IR, it seems like the `ensure` just trivially holds, but you replace `%z` with `%1` when proving the inside. If we'd separate the contract, the 'ensure' would use `%1` always, but you'd inline the alternative representation and thus the DCE is also trivial. On the negative side, you'd need to duplicate the `mul` and constant, but we could also factor out bigger bodies into a func or similar.

File: docs/Dialects/Verif.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (maerhart) [reply docs/Dialects/Verif.md:157]: I think the contract op could almost stay as is. Just remove the operands and results.
For the op that provides the alternative, I'm not exactly sure. If we'd specify contracts on the module level, I'd say, just put it in another module and link them with some `verif.alternative` op. But it'd also work to use the same structure as the contract has now, just without the `ensure` and `require` ops inside.

I'm generally also wondering if the contract doesn't have to specify more explicitly what part of the module it applies to. E.g., if you have a memory write operation. It wouldn't show up in a DFG when you just consider the fan-in of some SSA value. Similarly, for things like `llhd.drv` or anything that uses connect or alloca semantics in some way. I'm wondering how bad it would actually be to force a contract to apply to exactly one whole module? It doesn't necessarily have to be a module at the Chisel level, just at the HW level. The compiler can just carve out the part of a module the contract applies to and factor it out into its own private submodule.

```mlir
// Applies to the module implementation, and also every `verif.alternative` by
// inlining the alternative body into the module the way you currently inline contracts.
// What if there are multiple contracts and/or multiple alternatives?
// I think forcing them to operate on a per-module basis would simplify things greatly.
// You'd have 0 or 1 contract per module and any number of alternative
// implementations per module that cannot overlap.
verif.contract {
  %inputSum = comb.add %a0, %a1, %a2 : i42
  %outputSum = comb.add %z0, %z1 : i42
  verif.ensure_equal %inputSum, %outputSum : i42
}
%z0, %z1 = verif.alternative %0, %5 {
  %inputSum = comb.add %a0, %a1, %a2 : i42
  %any0 = verif.symbolic_value : i42
  %any1 = verif.symbolic_value : i42
  %anySum = comb.add %any0, %any1 : i42
  verif.assume_equal %anySum, %inputSum : i42
  verif.yield %any0, %any1 : i42, i42
}
```

Or a module-based design:
```mlir
module {
  // Written by the verification engineer or auto-generated from some spec. Makes it easier to put it in a different input file than the implementation by just having some `verif.contract.extern`. Design engineers can then implement several versions of designs that implement the same contract.
  verif.contract @contract(in %a0: i42, in %a1: i42, in %a2: i42, out %z0: i42, out %z1: i42) {
    %inputSum = comb.add %a0, %a1, %a2 : i42
    %outputSum = comb.add %z0, %z1 : i42
    verif.ensure_equal %inputSum, %outputSum : i42
  }

  // Verification flows look up the contract and set up a 'verif.formal' like this:
  // verif.formal {
  //   %a0 = verif.symbolic_value : i42
  //   %a1 = verif.symbolic_value : i42
  //   %a2 = verif.symbolic_value : i42
  //   %z0, %z1 = hw.instance @CarrySaveCompress3to2(%a0: i42, %a1: i42, %a2: i42) -> (z0: i42, z1: i42)
  //   %inputSum = comb.add %a0, %a1, %a2 : i42
  //   %outputSum = comb.add %z0, %z1 : i42
  //   verif.assert_equal %inputSum, %outputSum : i42
  // }
  hw.module @CarrySaveCompress3to2(in %a0: i42, in %a1: i42, in %a2: i42, out z0: i42, out z1: i42) implements @contract {
    %c1_i42 = hw.constant 1 : i42
    %0 = comb.xor %a0, %a1, %a2 : i42  // sum bits of FA (a0^a1^a2)
    %1 = comb.and %a0, %a1 : i42
    %2 = comb.or %a0, %a1 : i42
    %3 = comb.and %2, %a2 : i42
    %4 = comb.or %1, %3 : i42          // carry bits of FA (a0&a1 | a2&(a0|a1))
    %5 = comb.shl %4, %c1_i42 : i42    // %5 = carry << 1
     hw.output %0, %5 : i42, i42
  }

  // same for this alternative implementation
  hw.module @alt1(in %a0: i42, in %a1: i42, in %a2: i42, out z0: i42, out z1: i42) implements @contract {
    %inputSum = comb.add %a0, %a1, %a2 : i42
    %any0 = verif.symbolic_value : i42
    %any1 = verif.symbolic_value : i42
    %anySum = comb.add %any0, %any1 : i42
    verif.assume_equal %anySum, %inputSum : i42
    verif.yield %any0, %any1 : i42, i42
  }

  hw.module @top(in %a0: i42, in %a1: i42, in %a2: i42, out z0: i42, out z1: i42) {
    // Regular ExportVerilog/Synthesis flows just treat it as a regular instance and ignore the 'implements' attribute
    // Formal verification flows look up the module, see the 'implements' attribute, look up the 'contract', and inline it here by inserting symbolic values for all instance outputs and replacing 'ensure' -> 'assume'; 'require' -> 'assert'
    // It's the contract's job to constrain these symbolic values enough that the contract this @top module is implementing can be verified.
    %z0, %z1 = hw.instance @CarrySaveCompress3to2(%a0: i42, %a1: i42, %a2: i42) -> (z0: i42, z1: i42)
    // This results in:
    // %inputSum = comb.add %a0, %a1, %a2 : i42
    // %z0 = verif.symbolic_value : i42
    // %z1 = verif.symbolic_value : i42
    // %outputSum = comb.add %z0, %z1 : i42
    // verif.assume_equal %inputSum, %outputSum : i42
    hw.output %z0, %z1 : i42, i42
  }
  // Or, use the alternative implementation instead (need to iterate over all modules and check if it implements the same contract; maybe we can add another hint to indicate for what the alternative is intended because it could also be a golden model, symbolic implementation, simulation optimized implementation, etc.):
  hw.module @top(in %a0: i42, in %a1: i42, in %a2: i42, out z0: i42, out z1: i42) {
    %z0, %z1 = hw.instance @alt1(%a0: i42, %a1: i42, %a2: i42) -> (z0: i42, z1: i42)
    hw.output %z0, %z1 : i42, i42
  }
}
```

**Comment by fabianschuiki** [comment]
Thanks for the fantastic feedback @maerhart. I have tweaked the examples to reflect what we have discussed. Basically, the contracts now always replace their results with symbolic values when they are assumed to hold, which then makes the body of the contract constrain them appropriately. A simple canonicalization from `assume(equal(symbolic, x))` to a replacement of `symbolic -> x` can then get rid of the symbolic values in all cases where the contract provides an exact replacement for a value.

**Review by maerhart — APPROVED** [review]
Really nice that we were able to get rid of these alternative implementations and automatically use the contract itself instead. This makes these contracts much more solid, easier to work with, and reduces the overall tests we have to run because we don't have to check the alternatives as well!

We should probably update the contract op description in the tablegen file as well at some point to reflect this change.

File: docs/Dialects/Verif.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (dobios) [reply docs/Dialects/Verif.md:322]: a typical example that uses require would be like an implementation of an extract / subsequence operation that requires that the `min` operand is actually smaller than the `max` operand.

File: docs/Dialects/Verif.md [hunk 5]
```diff
@@ -4,6 +4,322 @@ This dialect provides a collection of operations to express various verification
...
+
+These contracts can then be used in two key ways:
+
+1.  A contract can be _verified_ to hold by turning `require`s into `assume`s and `ensure`s into `assert`s.
+    Doing so verifies that a circuit upholds the contract by placing asserts on the values it produces, and by placing assumes on the input values the circuit sees.
+    In a nutshell, this checks that, assuming the inputs honor the contract, the circuit also upholds the contract.
+    This check can be done very efficiently by creating `verif.formal` ops to verify each contract.
+
+2.  A contract can be _assumed_ to hold by turning `require`s into `assert`s and `ensure`s into `assume`s.
+    Doing so verifies that the inputs fed into a circuit uphold the contract, such that the outputs can be assumed to have the promised values.
```
Comment (dobios) [inline docs/Dialects/Verif.md:22]: I'm not entirely sure I'm a fan of this language. We're not assuming a contract, rather we're using it as an abstraction for the module during the verification of an instance.

File: docs/Dialects/Verif.md [hunk 6]
```diff
@@ -4,6 +4,322 @@ This dialect provides a collection of operations to express various verification
...
+## Contracts
+
+Formal contracts are a key building block of the Verif dialect to help make formal verification scale to larger designs and deep module hierarchies.
+Contracts describe what a circuit expects from its inputs (`verif.require`) and what it guarantees its output to be (`verif.ensure`).
+The `verif.contract` op can be inserted into an SSA edge similar to an `hw.wire`, where the contract simply passes its operands through to its results.
+During formal verification the contract results are replaced by symbolic values that uphold the guarantees described in the contract's body.
+
+These contracts can then be used in two key ways:
+
+1.  A contract can be _verified_ to hold by turning `require`s into `assume`s and `ensure`s into `assert`s.
```
Comment (dobios) [inline docs/Dialects/Verif.md:16]: I guess here you're not really verifying the contract, but you're checking that your region satisfies the contract, or that your region is valid under the contract. This follows the wording from the original paper this is based on.

File: docs/Dialects/Verif.md [hunk 7]
```diff
@@ -4,6 +4,322 @@ This dialect provides a collection of operations to express various verification
...
+    Doing so verifies that a circuit upholds the contract by placing asserts on the values it produces, and by placing assumes on the input values the circuit sees.
+    In a nutshell, this checks that, assuming the inputs honor the contract, the circuit also upholds the contract.
+    This check can be done very efficiently by creating `verif.formal` ops to verify each contract.
+
+2.  A contract can be _assumed_ to hold by turning `require`s into `assert`s and `ensure`s into `assume`s.
+    Doing so verifies that the inputs fed into a circuit uphold the contract, such that the outputs can be assumed to have the promised values.
+    In a nutshell, this checks that the inputs to a circuit honor the contract and therefore the circuit can be assumed to uphold the contract.
```
Comment (dobios) [inline docs/Dialects/Verif.md:23]: Again, just linking to the original work this is based on would help substantiate this claim.

File: docs/Dialects/Verif.md [hunk 8]
```diff
@@ -4,6 +4,322 @@ This dialect provides a collection of operations to express various verification
...
+  %1 = comb.add %a, %0 : i42
+
+  // Contract inlined with ensure -> assert, %z -> %1
+  %c9_i42 = hw.constant 9 : i42
+  %a9 = comb.mul %a, %c9_i42 : i42
+  verif.assert_equal %1, %a9 : i42
+}
+```
+
+Once the contract is proven, it can be assumed to hold everywhere.
```
Comment (dobios) [inline docs/Dialects/Verif.md:71]: I honestly prefer the term "checked" rather than proven, because we aren't actually proving anything, we're just doing a heuristic-based search of the input space for a satisfiability problem. Proving would require something like an inductive proof, which we don't actually do ^^"

File: docs/Dialects/Verif.md [hunk 9]
```diff
@@ -4,6 +4,322 @@ This dialect provides a collection of operations to express various verification
...
+Assuming it holds can be done by inlining it into its parent block and replacing its results with symbolic values.
+At the same time, `ensure`s are replaced with `assume`s and `require`s with `assert`s.
+Inlining the contract in this example and replacing the symbolic value with what it is assumed to be equal to will make the module produce the simple `9*a` term.
+This means that the ops describing the original implementation become obsolete and will be DCEd.
```
Comment (dobios) [inline docs/Dialects/Verif.md:75]: I hope this is true ^^"

**Review by dobios — APPROVED** [review]
This looks really great! thank you so much for writing this out!! I think I would also suggest adding an example that has a precondition, like @maerhart mentioned. Secondly, I would try to tie this in some part to the actual theoretical concept this is implementing. In its current state (so not tied to any module but rather to an arbitrary set of SSA values) a contract is basically an implementation of a hoare triple and our replacement (with the lowering to btor) encodes a weakest precondition computation like in the original 1975 [guarded commands paper](https://dl.acm.org/doi/10.1145/360933.360975) (since we don't have loops in our setting), that would also maybe give people a pointer to where they can find a proof that this actually verifies the design correctly.  This is of course a version that's adapted to hardware, but the idea is exactly the same.

Thanks again for doing this! This is great!!

**Comment by dobios** [comment]
also curious could someone point me to when `ensures_eq` was introduced, it seems to be a fairly redundant op, which will make lowering more annoying imo

File: docs/Dialects/Verif.md (hunk 5 above)
Comment (fabianschuiki) [reply docs/Dialects/Verif.md:22]: Yeah good point. I was trying to make the contract feel like an equivalent to asserts and assumes, where you just assert and assume a higher-level concept instead of individual signals. Especially since we don't have any modules and instances tied to the contracts. It feels more like a local abstraction that you can formulate which then gets substituted in.

File: docs/Dialects/Verif.md (hunk 8 above)
Comment (fabianschuiki) [reply docs/Dialects/Verif.md:71]: We could though :smirk: But yeah, you're right.

**Comment by fabianschuiki** [comment]
> @dobios also curious could someone point me to when ensures_eq was introduced, it seems to be a fairly redundant op, which will make lowering more annoying imo

It doesn't exist :grin: I just used it in the docs to not have to write the redundant `icmp eq` everywhere, with the additional SSA value. But in practice we'll just stick with the `ensure` and `require` we have, and the condition you can feed them.

**Comment by fabianschuiki** [comment]
Thanks for the great feedback @maerhart and @dobios. I'm landing this now with some more tweaks to the wording based on your feedback. Happy to iterate on this further as we go and get an implementation of this going.
